django/db/models/sql/where.py

```diff
@@ -35,48 +35,80 @@ class WhereNode(tree.Node):
     resolved = False
     conditional = True
 
-    def split_having(self, negated=False):
+    def must_remain_connected(self, negated=False):
         """
-        Return two possibly None nodes: one for those parts of self that
-        should be included in the WHERE clause and one for those parts of
-        self that must be included in the HAVING clause.
+        Whether or not children must be connected in the same filtering clause
+        (WHERE > HAVING > QUALIFY) to maintain logical semantic.
         """
-        if not self.contains_aggregate:
-            return self, None
         in_negated = negated ^ self.negated
-        # If the effective connector is OR or XOR and this node contains an
-        # aggregate, then we need to push the whole branch to HAVING clause.
-        may_need_split = (
+        return (
             (in_negated and self.connector == AND)
             or (not in_negated and self.connector == OR)
             or self.connector == XOR
         )
-        if may_need_split and self.contains_aggregate:
-            return None, self
+
+    def split_having_qualify(self, negated=False):
```

Comment:
Felt like `split()` could be a good enough name as `split_having_qualify` is pretty heavy.